File: duplicate-orders-from-a-double-submit/python/find_duplicate_orders.py
"""Find and cancel duplicate BigCommerce orders created by a double submit.

A slow payment gateway or an impatient double click on Place Order can turn one
checkout into two separate orders: same customer, same products, same total,
created seconds apart. This lists recent pre-fulfillment orders, groups them
with a pure function, keeps the earliest order in each group, and cancels the
rest, but only after re-checking that the duplicate has no captured payment.
Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/bigcommerce/duplicate-orders-from-a-double-submit/
"""
import os
import logging
from datetime import datetime, timedelta, timezone
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("find_duplicate_orders")

STORE_HASH = os.environ["BIGCOMMERCE_STORE_HASH"]
TOKEN = os.environ["BIGCOMMERCE_ACCESS_TOKEN"]
BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/"
WINDOW_SECONDS = int(os.environ.get("DUPLICATE_WINDOW_SECONDS", "300"))
LOOKBACK_MINUTES = int(os.environ.get("LOOKBACK_MINUTES", "15"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

HEADERS = {"X-Auth-Token": TOKEN, "Accept": "application/json", "Content-Type": "application/json"}

PRE_FULFILLMENT_STATUS_IDS = {0, 1, 7, 9, 11}
CANCELLED_STATUS_ID = 5


def bc_get(path, params=None):
    r = requests.get(BASE + path, headers=HEADERS, params=params or {}, timeout=30)
    r.raise_for_status()
    return r.json()


def bc_put(path, body):
    r = requests.put(BASE + path, headers=HEADERS, json=body, timeout=30)
    r.raise_for_status()
    return r.json()


def product_signature(products):
    """A stable string for a cart's contents, order does not matter."""
    parts = sorted(f"{p['product_id']}x{p['quantity']}" for p in products)
    return "|".join(parts)


def _parse(date_created):
    return datetime.strptime(date_created, "%a, %d %b %Y %H:%M:%S %z")


def find_duplicate_order_groups(orders, window_seconds=300):
    """Pure grouping and clustering logic. No I/O, no network.

    orders: list of {id, customer_id, date_created, total_inc_tax, status_id, product_signature}
    window_seconds: max gap between consecutive orders to still be one cluster
    returns: list of [keeper_id, *duplicate_ids], keeper is the earliest order in the cluster
    """
    eligible = [o for o in orders if o["status_id"] in PRE_FULFILLMENT_STATUS_IDS]

    groups = {}
    for o in eligible:
        key = (o["customer_id"], o["product_signature"], o["total_inc_tax"])
        groups.setdefault(key, []).append(o)

    duplicate_groups = []
    for members in groups.values():
        members = sorted(members, key=lambda o: _parse(o["date_created"]))
        cluster = [members[0]]
        for prev, curr in zip(members, members[1:]):
            delta = (_parse(curr["date_created"]) - _parse(prev["date_created"])).total_seconds()
            if delta <= window_seconds:
                cluster.append(curr)
            else:
                if len(cluster) > 1:
                    duplicate_groups.append([o["id"] for o in cluster])
                cluster = [curr]
        if len(cluster) > 1:
            duplicate_groups.append([o["id"] for o in cluster])

    return duplicate_groups


def recent_candidate_orders(min_date_created):
    orders = bc_get("v2/orders", {
        "min_date_created": min_date_created,
        "sort": "date_created:asc",
    })
    out = []
    for o in orders:
        if o["status_id"] not in PRE_FULFILLMENT_STATUS_IDS:
            continue
        products = bc_get(f"v2/orders/{o['id']}/products")
        out.append({
            "id": o["id"],
            "customer_id": o["customer_id"],
            "date_created": o["date_created"],
            "total_inc_tax": o["total_inc_tax"],
            "status_id": o["status_id"],
            "product_signature": product_signature(products),
        })
    return out


def has_settled_transaction(order_id):
    transactions = bc_get(f"v2/orders/{order_id}/transactions")
    return any(t.get("status") in ("captured", "authorized") for t in transactions)


def cancel_order(order_id):
    return bc_put(f"v2/orders/{order_id}", {"status_id": CANCELLED_STATUS_ID})


def run():
    min_date_created = (datetime.now(timezone.utc) - timedelta(minutes=LOOKBACK_MINUTES)).strftime(
        "%Y-%m-%dT%H:%M:%S+00:00"
    )
    orders = recent_candidate_orders(min_date_created)
    groups = find_duplicate_order_groups(orders, WINDOW_SECONDS)

    cancelled = 0
    flagged = 0
    for group in groups:
        keeper_id, *duplicate_ids = group
        for order_id in duplicate_ids:
            if has_settled_transaction(order_id):
                log.warning("Order %s has a captured transaction, flagging for manual refund then cancel.", order_id)
                flagged += 1
                continue
            log.info("Order %s is a duplicate of %s. %s", order_id, keeper_id,
                      "would cancel" if DRY_RUN else "cancelling")
            if not DRY_RUN:
                cancel_order(order_id)
            cancelled += 1

    log.info("Done. %d order(s) %s, %d flagged for manual review.",
              cancelled, "to cancel" if DRY_RUN else "cancelled", flagged)


if __name__ == "__main__":
    run()

File: duplicate-orders-from-a-double-submit/node/find-duplicate-orders.js
/**
 * Find and cancel duplicate BigCommerce orders created by a double submit.
 *
 * A slow payment gateway or an impatient double click on Place Order can turn one
 * checkout into two separate orders: same customer, same products, same total,
 * created seconds apart. This lists recent pre-fulfillment orders, groups them
 * with a pure function, keeps the earliest order in each group, and cancels the
 * rest, but only after re-checking that the duplicate has no captured payment.
 * Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/duplicate-orders-from-a-double-submit/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "storehash123";
const TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "dummy_token";
const BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/`;
const WINDOW_SECONDS = Number(process.env.DUPLICATE_WINDOW_SECONDS || 300);
const LOOKBACK_MINUTES = Number(process.env.LOOKBACK_MINUTES || 15);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const HEADERS = { "X-Auth-Token": TOKEN, "Accept": "application/json", "Content-Type": "application/json" };

const PRE_FULFILLMENT_STATUS_IDS = new Set([0, 1, 7, 9, 11]);
const CANCELLED_STATUS_ID = 5;

export function productSignature(products) {
  const parts = products.map((p) => `${p.product_id}x${p.quantity}`).sort();
  return parts.join("|");
}

function toMillis(dateCreated) {
  return new Date(dateCreated).getTime();
}

export function findDuplicateOrderGroups(orders, windowSeconds = 300) {
  const eligible = orders.filter((o) => PRE_FULFILLMENT_STATUS_IDS.has(o.status_id));

  const groups = new Map();
  for (const o of eligible) {
    const key = `${o.customer_id}|${o.product_signature}|${o.total_inc_tax}`;
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(o);
  }

  const duplicateGroups = [];
  for (const members of groups.values()) {
    const sorted = [...members].sort((a, b) => toMillis(a.date_created) - toMillis(b.date_created));
    let cluster = [sorted[0]];
    for (let i = 1; i < sorted.length; i++) {
      const deltaSeconds = (toMillis(sorted[i].date_created) - toMillis(sorted[i - 1].date_created)) / 1000;
      if (deltaSeconds <= windowSeconds) {
        cluster.push(sorted[i]);
      } else {
        if (cluster.length > 1) duplicateGroups.push(cluster.map((o) => o.id));
        cluster = [sorted[i]];
      }
    }
    if (cluster.length > 1) duplicateGroups.push(cluster.map((o) => o.id));
  }

  return duplicateGroups;
}

async function bcGet(path, params = {}) {
  const qs = new URLSearchParams(params).toString();
  const res = await fetch(BASE + path + (qs ? `?${qs}` : ""), { headers: HEADERS });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  return res.json();
}

async function bcPut(path, body) {
  const res = await fetch(BASE + path, { method: "PUT", headers: HEADERS, body: JSON.stringify(body) });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  return res.json();
}

async function recentCandidateOrders(minDateCreated) {
  const orders = await bcGet("v2/orders", {
    min_date_created: minDateCreated,
    sort: "date_created:asc",
  });
  const out = [];
  for (const o of orders) {
    if (!PRE_FULFILLMENT_STATUS_IDS.has(o.status_id)) continue;
    const products = await bcGet(`v2/orders/${o.id}/products`);
    out.push({
      id: o.id,
      customer_id: o.customer_id,
      date_created: o.date_created,
      total_inc_tax: o.total_inc_tax,
      status_id: o.status_id,
      product_signature: productSignature(products),
    });
  }
  return out;
}

async function hasSettledTransaction(orderId) {
  const transactions = await bcGet(`v2/orders/${orderId}/transactions`);
  return transactions.some((t) => ["captured", "authorized"].includes(t.status));
}

async function cancelOrder(orderId) {
  return bcPut(`v2/orders/${orderId}`, { status_id: CANCELLED_STATUS_ID });
}

export async function run() {
  const minDateCreated = new Date(Date.now() - LOOKBACK_MINUTES * 60 * 1000).toISOString();
  const orders = await recentCandidateOrders(minDateCreated);
  const groups = findDuplicateOrderGroups(orders, WINDOW_SECONDS);

  let cancelled = 0;
  let flagged = 0;
  for (const group of groups) {
    const [keeperId, ...duplicateIds] = group;
    for (const orderId of duplicateIds) {
      if (await hasSettledTransaction(orderId)) {
        console.warn(`Order ${orderId} has a captured transaction, flagging for manual refund then cancel.`);
        flagged++;
        continue;
      }
      console.log(`Order ${orderId} is a duplicate of ${keeperId}. ${DRY_RUN ? "would cancel" : "cancelling"}`);
      if (!DRY_RUN) await cancelOrder(orderId);
      cancelled++;
    }
  }

  console.log(`Done. ${cancelled} order(s) ${DRY_RUN ? "to cancel" : "cancelled"}, ${flagged} flagged for manual review.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
